lib/elixir/lib/calendar/datetime.ex

```diff
@@ -1674,6 +1675,143 @@ defmodule DateTime do
     end
   end
 
+  @doc """
+  Shifts given `datetime` by `duration` according to its calendar.
+
+  Allowed units are: `:year`, `:month`, `:week`, `:day`, `:hour`, `:minute`, `:second`, `:microsecond`.
+
+  Shifting datetimes in time zones that observe "Daylight Saving Time" across
+  summer/winter time will add/remove one hour from the resulting datetime.
+  This ensures `shift/3` always returns a valid datetime.
+
+      dt = DateTime.new!(~D[2019-03-31], ~T[01:00:00], "Europe/Copenhagen")
+      DateTime.shift(dt, hour: 1)
+      #=> #DateTime<2019-03-31 03:00:00+02:00 CEST Europe/Copenhagen>
+
+      dt = DateTime.new!(~D[2018-11-04], ~T[00:00:00], "America/Los_Angeles")
+      DateTime.shift(dt, hour: 2)
+      #=> #DateTime<2018-11-04 01:00:00-08:00 PST America/Los_Angeles>
```

Comment:
```suggestion
  This operation is equivalent to shifting the datetime wall clock (in other words,
  the values as we see them printed), then applying the time zone offset before
  computing the new time zone. This ensures `shift/3` always returns a valid
  datetime.

  On other other hand, time zones that observe "Daylight Saving Time"
  or other changes, across summer/winter time will add/remove hours
  from the resulting datetime:

      dt = DateTime.new!(~D[2019-03-31], ~T[01:00:00], "Europe/Copenhagen")
      DateTime.shift(dt, hour: 1)
      #=> #DateTime<2019-03-31 03:00:00+02:00 CEST Europe/Copenhagen>

      dt = DateTime.new!(~D[2018-11-04], ~T[00:00:00], "America/Los_Angeles")
      DateTime.shift(dt, hour: 2)
      #=> #DateTime<2018-11-04 01:00:00-08:00 PST America/Los_Angeles>

  In case you don't want these changes to happen automatically or you
  want to surface timezone conflicts to the user, you can shift
  the datetime as a naive datetime and then use `from_naive/2`:

      dt |> NaiveDateTime.shift(duration) |> DateTime.from_naive(dt.time_zone)
```